### PR TITLE
Handle string callable in case of non-static method

### PR DESCRIPTION
**Pull request by Jean85**
This is an improvement over #815. Thanks to @dmitry-pro for spotting the bug to start with.

### CODE REVIEW

**Comment by Jean85** [comment]
Umh, there's something strange... I cannot reproduce the bug outside of Sentry's tests: https://3v4l.org/1CEnC

What's really triggering the issue in here?

**Comment by dmitry-pro** [comment]
@Jean85 don't have much time for testing, but it is 
https://3v4l.org/XMiVr
Strange that it doesn't trigger in Your (previous) example - when I caught this error, it was found in public dynamic method, not function.
I'll test it more when I free.

**Comment by Jean85** [comment]
I've found the difference: the issue is present with functions, not with class methods: https://3v4l.org/4ibt7

Not sure why there's a difference, the bug seems to be very subtle.

**Comment by ste93cry** [comment]
Unfortunately we cannot remove the typehint or we will have a BC

**Comment by Jean85** [comment]
> Unfortunately we cannot remove the typehint or we will have a BC

Crap that's true... maybe we can hotfix it in the previous method, but it would still be broken for other usages

**Comment by chekalsky** [comment]
@Jean85 When are you planning to release it? Pretty annoying bug :-)

File: src/Serializer/AbstractSerializer.php [hunk 1]
```diff
@@ -261,17 +261,44 @@ protected function serializeValue($value)
...
+
+        if (\is_string($callable) && !\function_exists($callable)) {
+            return $callable;
+        }
+
+        return $this->serializeCallable($callable);
+    }
+
+    /**
+     * @internal Use serializeCallableWithoutTypeHint instead (no type in argument)
```
Comment (ste93cry) [inline src/Serializer/AbstractSerializer.php]: We cannot mark an already existing method as `@internal` or we may be tempted in the future to make changes without worrying about BC when we must instead since it was part of the public API in the past. Just leave the note instead with a link to the issue and remove the comment from the parameter instead

File: src/Serializer/AbstractSerializer.php [hunk 2]
```diff
@@ -261,17 +261,44 @@ protected function serializeValue($value)
...
+     * since using the callable type raises a deprecation in some cases.
+     *
+     * @param callable $callable
+     *
+     * @return string
+     */
+    protected function serializeCallableWithoutTypeHint($callable): string
+    {
+        if (!\is_callable($callable)) {
+            throw new \InvalidArgumentException('Expecting callable, got ' . \gettype($callable));
```
Comment (ste93cry) [inline src/Serializer/AbstractSerializer.php]: Please use `sprintf` and handle the fact that `$callable` may be either an object or a scalar

File: src/Serializer/AbstractSerializer.php [hunk 3]
```diff
@@ -261,17 +261,44 @@ protected function serializeValue($value)
...
+     *
      * @param callable $callable
      *
      * @return string
      */
-    protected function serializeCallable(callable $callable): string
+    protected function serializeCallable(/* callable type to be removed in 3.0, see #821 => */callable $callable): string
     {
+        if (!\is_callable($callable)) {
+            throw new \InvalidArgumentException('Expecting callable, got ' . \gettype($callable));
```
Comment (ste93cry) [inline src/Serializer/AbstractSerializer.php]: Please use `sprintf` and handle the fact that `$callable` may be either an object or a scalar

File: src/Serializer/AbstractSerializer.php [hunk 4]
```diff
@@ -357,4 +383,21 @@ public function getSerializeAllObjects(): bool
     {
         return $this->serializeAllObjects;
     }
+
+    /**
+     * @param mixed $callable
+     *
+     * @throws InvalidArgumentException
+     */
+    private function assertIsValidCallable($callable): void
```
Comment (ste93cry) [inline src/Serializer/AbstractSerializer.php]: I don't see any good reason to have (in this case) a `private` method that just checks whether the argument is a callable and throws if it's not. Please inline the code where it belongs to

File: src/Serializer/AbstractSerializer.php [hunk 5]
```diff
@@ -261,17 +262,42 @@ protected function serializeValue($value)
...
+     * Use serializeCallableWithoutTypeHint instead (no type in argument).
+     *
+     * @see https://github.com/getsentry/sentry-php/pull/821
+     *
      * @param callable $callable
      *
      * @return string
      */
-    protected function serializeCallable(callable $callable): string
+    protected function serializeCallable(/* callable type to be removed in 3.0, see #821 => */callable $callable): string
```
Comment (ste93cry) [inline src/Serializer/AbstractSerializer.php]: Instead of putting this comment here use the description of the method

File: composer.json [hunk 6]
```diff
@@ -74,7 +74,7 @@
     },
     "extra": {
         "branch-alias": {
-            "dev-develop": "2.2-dev"
+            "dev-develop": "2.3-dev"
```
Comment (ste93cry) [inline composer.json]: Please revert this change as this PR is targeting `master`

**Comment by emircanerkul** [comment]
Hi @ste93cry @Jean85 

I face the same issue now while using sentry/sentry 3.22.1.

Is there any regression?

```
Deprecated: Use of "static" in callables is deprecated in /app/vendor/sentry/sentry/src/Serializer/AbstractSerializer.php on line 265 Deprecated: Use of "static" in callables is deprecated in /app/vendor/sentry/sentry/src/Serializer/AbstractSerializer.php on line 265 Deprecated: Use of "static" in callables is deprecated in /app/vendor/sentry/sentry/src/Serializer/AbstractSerializer.php on line 265 Deprecated: Use of "static" in callables is deprecated in /app/vendor/sentry/sentry/src/Serializer/AbstractSerializer.php on line 265 Deprecated: Use of "static" in callables is deprecated in /app/vendor/sentry/sentry/src/Serializer/AbstractSerializer.php on line 265 Deprecated: Use of "static" in callables is deprecated in /app/vendor/sentry/sentry/src/Serializer/AbstractSerializer.php on line 265
```

**Comment by ste93cry** [comment]
It may be, even though it's the first time it is reported since years... I suggest to try version `4.x` of the SDK, and if the issue persist open a new issue.
